torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -416,6 +425,10 @@ def _pre_forward(
         # their gradients. They must be re-registered every forward pass in case
         # the `grad_fn` is mutated.
         _register_post_backward_hooks(state, handles)
+        # We may have to reallocate the _cpu_grad if optimizer overlap set the grad to None in the backward pass.
```

Comment:
Is this true? If we set `.grad` to `None`, the `._cpu_grad` still exists?

Reply:
@awgu I'm keeping this since I'm setting cpu_grad=None in the post-backward. This is to maintain the invariant that if all orig param grad = None, then handle.sharded_grad should be None. In this case (when CPU offload enabled), handle.sharded_grad == cpu_grad.

Reply:
I feel like the invariant is valuable because one of the goals of optimizer in backward is to cut down on memory usage by not holding on to sharded gradients that have already been applied on the parameter. 

Due to this, in GPU case, it seems essential that we set `handle.sharded_grad` (whether sharded_grad is the flat_param.grad or flat_param._saved_grad_shard) to None - if we don't do this and just set `orig_param.grad = None`, we won't see a memory reduction since FSDP internals will still hold on to a grad. 

in CPU case, it is a little trickier - ideally we'd like to set `_cpu_grad = None` as well, but this results in the extra reallocation in the next forward. Since we're in CPU memory, saving memory is less crucial because usually GPU memory << CPU memory, so maybe we just zero the cpu_grad instead of setting to None.

Reply:
Thanks for this analysis! I am fully onboard for the GPU case, but I am still skeptical about the performance on the CPU case (and I do not think right now that the CPU memory savings from freeing sharded gradients is meaningful). I am worried that the cost for this invariant is too high. Maybe we can go with this invariant first, but we can look at the performance (e.g. via traces) and reevaluate?